# Incremental Capstone — Session 1

## Aura Healthcare Data Analysis

### Objective

The goal of this session is to begin exploring the NSMES1988 healthcare dataset.

In this notebook, we will:

- Import the required Python libraries
- Load the healthcare dataset
- Inspect the structure of the data
- Review the columns and data types
- Check for missing values
- Examine basic descriptive statistics
- Begin identifying areas that may require cleaning or transformation

## 1. Import Required Libraries

We will begin by importing the Python libraries needed for data analysis.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv(r"C:\Users\juanb\Documents\Python_Course_End_Projects\incremental_capstone\data\NSMES1988.csv")


## 2. Initial Data Inspection

Before cleaning or transforming the dataset, we will inspect its size, columns, data types, and overall structure.

In [3]:
df.head()

,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,1,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.8810,yes,yes,no
1,2,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.7478,no,yes,no
2,3,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.6532,no,no,yes
3,4,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.6588,no,yes,no
4,5,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.6588,no,yes,no


In [4]:
df.shape

(4406, 19)

In [5]:
df.columns

Index(['Unnamed: 0', 'visits', 'nvisits', 'ovisits', 'novisits', 'emergency',
       'hospital', 'health', 'chronic', 'adl', 'region', 'age', 'gender',
       'married', 'school', 'income', 'employed', 'insurance', 'medicaid'],
      dtype='str')

### Dataset Structure

The dataset contains 4,406 observations and 19 columns. An initial review of the column names shows an `Unnamed: 0` column, which may represent a previously saved index rather than an actual analytical variable. This column will be investigated before deciding whether to remove it.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   str    
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   str    
 10  region      4406 non-null   str    
 11  age         4406 non-null   float64
 12  gender      4406 non-null   str    
 13  married     4406 non-null   str    
 14  school      4406 non-null   int64  
 15  income      4406 non-null   float64
 16  employed    4406 non-null   str    
 17  insurance   4406 non-null   str    
 18  medicaid    4406 non-null   str    
dtypes: float64(2), int64(9), str(8)
memory

### Missing Value Analysis

Although `df.info()` indicates that all columns contain 4,406 non-null values, an explicit missing-value check will be performed to confirm that no values are missing from the dataset.

In [7]:
df.isna().sum()

Unnamed: 0    0
visits        0
nvisits       0
ovisits       0
novisits      0
emergency     0
hospital      0
health        0
chronic       0
adl           0
region        0
age           0
gender        0
married       0
school        0
income        0
employed      0
insurance     0
medicaid      0
dtype: int64

### Investigating the `Unnamed: 0` Column

The dataset contains a column named `Unnamed: 0`, which may represent an index that was previously saved to the CSV file. Before removing the column, its values will be inspected to determine whether it contains meaningful analytical information.

In [8]:
df["Unnamed: 0"].head(10)

0     1
1     2
2     3
3     4
4     5
5     6
6     7
7     8
8     9
9    10
Name: Unnamed: 0, dtype: int64

In [9]:
df.shape

(4406, 19)

### Age and Income Analysis

The `age` and `income` columns will be examined using descriptive statistics to understand their values and ranges. This analysis includes the count, mean, standard deviation, minimum, maximum, and quartile values for each variable.

In [10]:
df[["age", "income"]].describe()

,age,income
count,4406.000000,4406.000000
mean,7.402406,2.527132
std,0.633405,2.924648
min,6.600000,-1.012500
25%,6.900000,0.912150
50%,7.300000,1.698150
75%,7.800000,3.172850
max,10.900000,54.835100


### Age and Income Observations

Both `age` and `income` are stored as `float64` numerical variables, and neither column contains missing values.

The `age` values range from 6.6 to 10.9, with a mean of approximately 7.40. According to the dataset description, age is represented in years divided by 10. Therefore, the actual age range is approximately 66 to 109 years, with an average age of approximately 74 years.

The `income` values range from -1.0125 to 54.8351, with a mean of approximately 2.53. Income is represented in units of $10,000. Therefore, the values correspond to an approximate range of -$10,125 to $548,351, with an average income of approximately $25,271.

The negative minimum income value should not automatically be treated as an error without additional information about how income was defined in the original dataset.

In [11]:
df.to_json("C:\\Users\\juanb\\Documents\\Python_Course_End_Projects\\incremental_capstone\\data\\NSMES1988.json", orient="records", indent=4)


## Memory Usage Analysis

The DataFrame will be examined to determine how much memory it uses and whether any columns could be stored using more efficient data types.

In [12]:
df.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  4406 non-null   int64  
 1   visits      4406 non-null   int64  
 2   nvisits     4406 non-null   int64  
 3   ovisits     4406 non-null   int64  
 4   novisits    4406 non-null   int64  
 5   emergency   4406 non-null   int64  
 6   hospital    4406 non-null   int64  
 7   health      4406 non-null   str    
 8   chronic     4406 non-null   int64  
 9   adl         4406 non-null   str    
 10  region      4406 non-null   str    
 11  age         4406 non-null   float64
 12  gender      4406 non-null   str    
 13  married     4406 non-null   str    
 14  school      4406 non-null   int64  
 15  income      4406 non-null   float64
 16  employed    4406 non-null   str    
 17  insurance   4406 non-null   str    
 18  medicaid    4406 non-null   str    
dtypes: float64(2), int64(9), str(8)
memory

### Unique Values by Column

The number of unique values in each column will be examined to identify variables with a small number of repeated values. Columns containing repeated categorical values may be candidates for conversion to the Pandas `category` data type to reduce memory usage.

In [13]:
df.nunique()

Unnamed: 0    4406
visits          60
nvisits         51
ovisits         37
novisits        35
emergency       11
hospital         9
health           3
chronic          9
adl              2
region           4
age             36
gender           2
married          2
school          19
income        3015
employed         2
insurance        2
medicaid         2
dtype: int64

### Integer Value Ranges

The minimum and maximum values of the integer columns will be examined before recommending smaller integer data types. The range of each variable must fit safely within the limits of the proposed data type.

In [14]:
df.select_dtypes(include='int64').agg(['min', 'max'])

,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,chronic,school
min,1,0,0,0,0,0,0,0,0
max,4406,89,104,141,155,12,8,8,18


### Recommended Data Type Optimizations

The DataFrame uses approximately 2.4 MB of memory when measured using deep memory inspection.

Several string columns contain only a small number of unique repeated values. The `health`, `adl`, `region`, `gender`, `married`, `employed`, `insurance`, and `medicaid` columns are therefore good candidates for conversion from string data types to the Pandas `category` data type.

The integer variables representing visits, hospital use, chronic conditions, and years of schooling contain relatively small values. These columns could be stored using smaller integer data types, such as `int16`, rather than the default `int64`, which would reduce memory usage while safely representing their observed ranges.

The `age` and `income` columns are currently stored as `float64`. These should remain as floating-point variables at this stage to preserve their numerical precision. Their units will be addressed during subsequent data preparation.

The `Unnamed: 0` column appears to be a previously saved row index rather than an analytical variable. Rather than optimizing its data type, it should be considered for removal before detailed analysis.

### Test Memory Optimization on a Copy

A copy of the original DataFrame will be created so that alternative data types can be tested without modifying the original dataset. The optimized copy will then be compared with the original DataFrame to measure the potential memory savings.

In [15]:
df_optimized = df.copy()

In [16]:
categorical_columns = [
    "health",
    "adl",
    "region",
    "gender",
    "married",
    "employed",
    "insurance",
    "medicaid"
]

df_optimized[categorical_columns] = df_optimized[categorical_columns].astype("category")

In [17]:
df_optimized[categorical_columns].dtypes

health       category
adl          category
region       category
gender       category
married      category
employed     category
insurance    category
medicaid     category
dtype: object

In [18]:
integer_columns = [
    "visits",
    "nvisits",
    "ovisits",
    "novisits",
    "emergency",
    "hospital",
    "chronic",
    "school"
]

df_optimized[integer_columns] = df_optimized[integer_columns].astype("int16")

In [19]:
df_optimized[integer_columns].dtypes

visits       int16
nvisits      int16
ovisits      int16
novisits     int16
emergency    int16
hospital     int16
chronic      int16
school       int16
dtype: object

In [20]:
df_optimized.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   Unnamed: 0  4406 non-null   int64   
 1   visits      4406 non-null   int16   
 2   nvisits     4406 non-null   int16   
 3   ovisits     4406 non-null   int16   
 4   novisits    4406 non-null   int16   
 5   emergency   4406 non-null   int16   
 6   hospital    4406 non-null   int16   
 7   health      4406 non-null   category
 8   chronic     4406 non-null   int16   
 9   adl         4406 non-null   category
 10  region      4406 non-null   category
 11  age         4406 non-null   float64 
 12  gender      4406 non-null   category
 13  married     4406 non-null   category
 14  school      4406 non-null   int16   
 15  income      4406 non-null   float64 
 16  employed    4406 non-null   category
 17  insurance   4406 non-null   category
 18  medicaid    4406 non-null   category
dtypes: category(8), f

### Memory Optimization Results

The original DataFrame used approximately 2.4 MB of memory when measured using deep memory inspection. After testing more appropriate data types, the optimized DataFrame uses approximately 207.8 KB.

The `health`, `adl`, `region`, `gender`, `married`, `employed`, `insurance`, and `medicaid` variables contain only a small number of repeated values and were converted from strings to the Pandas `category` data type.

The `visits`, `nvisits`, `ovisits`, `novisits`, `emergency`, `hospital`, `chronic`, and `school` variables contain relatively small whole-number values and were converted from `int64` to `int16`.

The `age` and `income` variables remain `float64` to preserve their numerical precision. These changes reduced memory usage substantially while maintaining the information contained in the dataset.

## Recommended Changes Before Detailed Analysis

Based on the initial inspection and memory analysis, several changes are recommended before performing detailed analysis:

- Remove the `Unnamed: 0` column because it contains sequential row numbers and appears to be an index saved with the original CSV rather than a healthcare variable.
- Convert the categorical variables (`health`, `adl`, `region`, `gender`, `married`, `employed`, `insurance`, and `medicaid`) to the Pandas `category` data type because they contain a small number of repeated values.
- Store the small whole-number variables (`visits`, `nvisits`, `ovisits`, `novisits`, `emergency`, `hospital`, `chronic`, and `school`) using a smaller integer data type such as `int16`.
- The `age` variable is currently represented as age divided by 10 and should eventually be converted to actual years before detailed analysis.
- The `income` variable is represented in units of $10,000 and should eventually be converted to actual dollar values before detailed analysis.
- The negative minimum income value should be investigated before deciding whether it represents an error or a valid value.
- No missing-value treatment is required because all 4,406 observations contain non-null values in every column.

In [21]:
df_optimized

,Unnamed: 0,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,1,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.881000,yes,yes,no
1,2,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.747800,no,yes,no
2,3,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.653200,no,no,yes
3,4,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.658800,no,yes,no
4,5,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.658800,no,yes,no
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4401,4402,11,0,0,0,0,0,average,0,normal,other,8.4,female,yes,8,2.249700,no,yes,no
4402,4403,12,0,0,0,0,0,average,2,normal,other,7.8,female,no,11,5.813200,no,yes,no
4403,4404,10,0,20,0,1,1,average,5,normal,other,7.3,male,yes,12,3.877916,no,yes,no
4404,4405,16,1,0,0,0,0,average,0,normal,other,6.6,female,yes,12,3.877916,no,yes,no


In [22]:
df_optimized = df_optimized.drop(columns=["Unnamed: 0"])

In [23]:
df_optimized.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 4406 entries, 0 to 4405
Data columns (total 18 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   visits     4406 non-null   int16   
 1   nvisits    4406 non-null   int16   
 2   ovisits    4406 non-null   int16   
 3   novisits   4406 non-null   int16   
 4   emergency  4406 non-null   int16   
 5   hospital   4406 non-null   int16   
 6   health     4406 non-null   category
 7   chronic    4406 non-null   int16   
 8   adl        4406 non-null   category
 9   region     4406 non-null   category
 10  age        4406 non-null   float64 
 11  gender     4406 non-null   category
 12  married    4406 non-null   category
 13  school     4406 non-null   int16   
 14  income     4406 non-null   float64 
 15  employed   4406 non-null   category
 16  insurance  4406 non-null   category
 17  medicaid   4406 non-null   category
dtypes: category(8), float64(2), int16(8)
memory usage: 173.6 KB


## Export Prepared DataFrame

The prepared DataFrame will be exported as `NSMES1988new.csv` for use in subsequent capstone assignments. The Pandas index will not be included as an additional column in the exported file.

In [27]:
df_optimized.to_csv("../data/NSMES1988new.csv", index=False)

In [28]:
df_new = pd.read_csv("../data/NSMES1988new.csv")

df_new.head()

,visits,nvisits,ovisits,novisits,emergency,hospital,health,chronic,adl,region,age,gender,married,school,income,employed,insurance,medicaid
0,5,0,0,0,0,1,average,2,normal,other,6.9,male,yes,6,2.8810,yes,yes,no
1,1,0,2,0,2,0,average,2,normal,other,7.4,female,yes,10,2.7478,no,yes,no
2,13,0,0,0,3,3,poor,4,limited,other,6.6,female,no,10,0.6532,no,no,yes
3,16,0,5,0,1,1,poor,2,limited,other,7.6,male,yes,3,0.6588,no,yes,no
4,3,0,0,0,0,0,average,2,limited,other,7.9,female,yes,6,0.6588,no,yes,no


In [26]:
df_new.shape

(4406, 18)

## Session 1 — Data Observation Report

The NSMES1988 healthcare dataset initially contained 4,406 observations and 19 columns. Inspection of the dataset showed that all columns contained 4,406 non-null values, indicating that no missing values were present.

The original dataset contained numerical and categorical variables. Eight variables (`health`, `adl`, `region`, `gender`, `married`, `employed`, `insurance`, and `medicaid`) contained a small number of repeated categorical values. These variables were identified as appropriate candidates for the Pandas `category` data type.

The numerical visit, hospital, chronic condition, and education variables contained relatively small whole-number ranges. Testing these variables as `int16` rather than the default `int64`, along with converting the categorical variables to `category`, substantially reduced DataFrame memory usage. Deep memory usage decreased from approximately 2.4 MB to 207.8 KB during the initial optimization test.

The `Unnamed: 0` column contained sequential values from 1 through 4,406 and appeared to represent a previously saved row index rather than a healthcare variable. It was removed from the prepared DataFrame. After its removal, the prepared dataset contained 4,406 observations and 18 analytical variables and used approximately 173.6 KB of memory.

The `age` variable ranges from 6.6 to 10.9 and represents age in years divided by 10. The `income` variable ranges from approximately -1.01 to 54.84 and represents family income in units of $10,000. These variables were left in their original units during Session 1 because their transformations will be performed during the next stage of data processing.

The prepared dataset was exported as `NSMES1988new.csv` for use in subsequent capstone sessions.